src/types.rs

```diff
@@ -0,0 +1,35 @@
+//! Concrete node and edge types used in the registry.
+
+/// The type of a node.
+pub enum NodeType {
+    /// A user, eg. contributor, project member etc.
+    User,
+    /// A project with users as members and contributors.
+    Project,
+}
+
+/// The type of an edge.
+pub enum EdgeType {
+    /// Contribution, user -> project, with strength metric.
+    ContributionToProject(u32),
+    /// Contribution, project -> user, with strength metric.
+    ContributionFromUser(u32),
```

Comment:
As the `strength metric` will be the same for both directions, this could be consolidated into a `project <-> user` edge.

On the other hand, I am not quite sure yet how to deal with bi-directional edges in the actual graph. We are currently using `petgraph` which to my knowledge does not support bi-directional edges. We were wondering already if we should find an alternative library or representation of the graph as a whole, but this needs more investigation.

Reply:
I see - so you would determine the edge weight based on where it originates from?

Reply:
I am not quite sure what you mean. A contribution edge is weighted by the number of contributions of this connection divided by the total number of contributions of the current node.

Reply:
If you only have one edge type for contributions, how do you know if it's a `user -> project` type edge or a `project -> user` one?

Reply:
In case of the random walker, the type of the current node is known so you that the edge type is indirectly known. For the algorithm itself the important informations to decide on the weight of an contribution type edge is `total number of contributions of the node` (which should be accessible no matter the type of the node and `number of contributions of the connection` from the edge. 
Not sure if I answered you question. If not, it would be great if you could add some more explanations.

Reply:
Ok, this is what I meant by:

> I see - so you would determine the edge weight based on where it originates from?

So we I'll change it to only have one Contribution edge type.